cranelift-codegen/meta/src/cdsl/instructions.rs

```diff
@@ -7,41 +14,51 @@ use crate::cdsl::type_inference::Constraint;
 use crate::cdsl::types::{LaneType, ValueType};
 use crate::cdsl::typevar::TypeVar;
 
-use std::fmt;
-use std::ops;
-use std::rc::Rc;
-use std::slice;
+#[derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct OpcodeNumber(u32);
+entity_impl!(OpcodeNumber);
+
+pub type AllInstructions = PrimaryMap<OpcodeNumber, Instruction>;
 
-pub struct InstructionGroupBuilder<'format_reg> {
+pub struct InstructionGroupBuilder<'format_reg, 'all_inst> {
     _name: &'static str,
     _doc: &'static str,
     format_registry: &'format_reg FormatRegistry,
-    instructions: Vec<Instruction>,
+    all_instructions: &'all_inst mut AllInstructions,
+    own_instructions: Vec<Instruction>,
 }
 
-impl<'format_reg> InstructionGroupBuilder<'format_reg> {
+impl<'format_reg, 'all_inst> InstructionGroupBuilder<'format_reg, 'all_inst> {
     pub fn new(
         name: &'static str,
         doc: &'static str,
+        all_instructions: &'all_inst mut AllInstructions,
         format_registry: &'format_reg FormatRegistry,
     ) -> Self {
         Self {
             _name: name,
             _doc: doc,
             format_registry,
-            instructions: Vec::new(),
+            all_instructions,
+            own_instructions: Vec::new(),
         }
     }
 
     pub fn push(&mut self, builder: InstructionBuilder) {
-        self.instructions.push(builder.build(self.format_registry));
+        // gen_inst wants the opcode number to start from 1.
+        // TODO this can change once we get rid of the Python code.
+        let opcode_number = OpcodeNumber(self.all_instructions.next_key().as_u32() + 1);
```

Comment:
Thats right, that's because gen_inst wants the opcode number to start from 1. You have a good point that it could be mistaken for the `all_instructions` index, which is not what it should mean. Probably instead I should just save the actual index as the opcode number, and offset it when generating the file.